// @anon/proxy-types/dist/proxyErrors.js
export class InvalidGeocodingError extends Error {
    ipAddress;
    latitude;
    longitude;
    constructor(ipAddress, latitude, longitude) {
        super(`Invalid geocoding for IP Address ${ipAddress} with Latitude: ${latitude}, Longitude: ${longitude}`);
        this.ipAddress = ipAddress;
        this.latitude = latitude;
        this.longitude = longitude;
        // Set the prototype explicitly.
        Object.setPrototypeOf(this, InvalidGeocodingError.prototype);
        // Ensure the name is set to the name of the new error
        this.name = this.constructor.name;
    }
}
export class InvalidArgumentError extends Error {
    inputValue;
    fieldName;
    functionName;
    constructor(inputValue, functionName, parameterName) {
        super(`Function (${functionName}) received invalid argument for ${parameterName}: ${inputValue}`);
        this.inputValue = inputValue;
        this.fieldName = parameterName;
        this.functionName = functionName;
        // Set the prototype explicitly.
        Object.setPrototypeOf(this, InvalidArgumentError.prototype);
        // Ensure the name is set to the name of the new error
        this.name = this.constructor.name;
    }
}
export class RequestClientError extends Error {
    service;
    path;
    status;
    errorDescription;
    constructor(service, path, status, errorDescription) {
        super(`Made a client error to ${service} (${path})  with : ${status} ${errorDescription}`);
        this.service = service;
        this.path = path;
        this.status = status;
        this.errorDescription = errorDescription;
        // Set the prototype explicitly.
        Object.setPrototypeOf(this, RequestClientError.prototype);
        // Ensure the name is set to the name of the new error
        this.name = this.constructor.name;
    }
}
export class RequestServerError extends Error {
    service;
    path;
    status;
    errorDescription;
    constructor(service, path, status, errorDescription) {
        super(`Made a server error to ${service} (${path})  with : ${status} ${errorDescription}`);
        this.service = service;
        this.path = path;
        this.status = status;
        this.errorDescription = errorDescription;
        // Set the prototype explicitly.
        Object.setPrototypeOf(this, RequestServerError.prototype);
    }
}
export class UnavailableProxyError extends Error {
    constructor(city, country) {
        super(`No proxies available around ${city}, ${country}`);
        // Set the prototype explicitly.
        Object.setPrototypeOf(this, UnavailableProxyError.prototype);
        // Ensure the name is set to the name of the new error
        this.name = this.constructor.name;
    }
}
