// @anon/fern-sdk/dist/serialization/resources/actions/types/LinkedinGetProfileResponse.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (
          !desc ||
          ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)
        ) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k];
            },
          };
        }
        Object.defineProperty(o, k2, desc);
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
      });
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
      }
    : function (o, v) {
        o["default"] = v;
      });
var __importStar =
  (this && this.__importStar) ||
  function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null)
      for (var k in mod)
        if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
          __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
  };
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinkedinGetProfileResponse = void 0;
const core = __importStar(require("../../../../core"));
const Education_1 = require("./Education");
const Language_1 = require("./Language");
const Skill_1 = require("./Skill");
const Certification_1 = require("./Certification");
const NetworkDistance_1 = require("./NetworkDistance");
const ContactInfo_1 = require("./ContactInfo");
const Birthdate_1 = require("./Birthdate");
const PrimaryLocale_1 = require("./PrimaryLocale");
const WorkExperience_1 = require("./WorkExperience");
const VolunteeringExperience_1 = require("./VolunteeringExperience");
exports.LinkedinGetProfileResponse = core.serialization.object({
  id: core.serialization.string(),
  firstName: core.serialization.string().optional(),
  lastName: core.serialization.string().optional(),
  headline: core.serialization.string(),
  profilePictureUrl: core.serialization.string().optional(),
  publicIdentifier: core.serialization.string().optional(),
  publicProfileUrl: core.serialization.string().optional(),
  location: core.serialization.string().optional(),
  summary: core.serialization.string().optional(),
  education: core.serialization.list(Education_1.Education).optional(),
  languages: core.serialization.list(Language_1.Language).optional(),
  skills: core.serialization.list(Skill_1.Skill).optional(),
  certifications: core.serialization
    .list(Certification_1.Certification)
    .optional(),
  networkDistance: NetworkDistance_1.NetworkDistance.optional(),
  contactInfo: ContactInfo_1.ContactInfo.optional(),
  websites: core.serialization.list(core.serialization.string()),
  birthdate: Birthdate_1.Birthdate.optional(),
  primaryLocale: PrimaryLocale_1.PrimaryLocale.optional(),
  backgroundPictureUrl: core.serialization.string().optional(),
  hashtags: core.serialization.list(core.serialization.string()).optional(),
  workExperience: core.serialization
    .list(WorkExperience_1.WorkExperience)
    .optional(),
  volunteeringExperience: core.serialization
    .list(VolunteeringExperience_1.VolunteeringExperience)
    .optional(),
  followerCount: core.serialization.number().optional(),
  connectionsCount: core.serialization.number().optional(),
  sharedConnectionsCount: core.serialization.number().optional(),
});
