// @anon/fern-sdk/dist/api/resources/userPool/client/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (
          !desc ||
          ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)
        ) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k];
            },
          };
        }
        Object.defineProperty(o, k2, desc);
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
      });
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
      }
    : function (o, v) {
        o["default"] = v;
      });
var __importStar =
  (this && this.__importStar) ||
  function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null)
      for (var k in mod)
        if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
          __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
  };
var __awaiter =
  (this && this.__awaiter) ||
  function (thisArg, _arguments, P, generator) {
    function adopt(value) {
      return value instanceof P
        ? value
        : new P(function (resolve) {
            resolve(value);
          });
    }
    return new (P || (P = Promise))(function (resolve, reject) {
      function fulfilled(value) {
        try {
          step(generator.next(value));
        } catch (e) {
          reject(e);
        }
      }
      function rejected(value) {
        try {
          step(generator["throw"](value));
        } catch (e) {
          reject(e);
        }
      }
      function step(result) {
        result.done
          ? resolve(result.value)
          : adopt(result.value).then(fulfilled, rejected);
      }
      step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
  };
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod };
  };
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserPool = void 0;
const environments = __importStar(require("../../../../environments"));
const core = __importStar(require("../../../../core"));
const serializers = __importStar(require("../../../../serialization/index"));
const url_join_1 = __importDefault(require("url-join"));
const errors = __importStar(require("../../../../errors/index"));
class UserPool {
  constructor(_options) {
    this._options = _options;
  }
  /**
   * This endpoint creates a new `UserPool`. UserPools associate a particular application with a group of users, and determines how those users are identified and authenticated via the `authorization` field. `authorization` should use the public JWKS URI of your identity provider. If you would like to use an `idClaim` other than the default `sub`, such as `email`, you may specify it here as well. If `authorization` is omitted, the UserPool will be configured as "Anon-hosted". This means you can use Anon as your identity provider by issuing AppUserIdTokens from `/org/appUserIdToken`.
   *
   * @param {AnonApi.CreateUserPool} request
   * @param {UserPool.RequestOptions} requestOptions - Request-specific configuration.
   *
   * @example
   *     await client.userPool.createUserPool({
   *         name: "string",
   *         description: "string",
   *         authorization: {
   *             jwksUri: "string",
   *             idClaim: "string"
   *         }
   *     })
   */
  createUserPool(request, requestOptions) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
      const _response = yield core.fetcher({
        url: (0, url_join_1.default)(
          (_a = yield core.Supplier.get(this._options.environment)) !== null &&
            _a !== void 0
            ? _a
            : environments.AnonApiEnvironment.Sandbox,
          "/org/userPool",
        ),
        method: "POST",
        headers: {
          Authorization: yield this._getAuthorizationHeader(),
          "X-Anon-Sdk-Client-Id": yield core.Supplier.get(
            this._options.anonSdkClientId,
          ),
          "X-Fern-Language": "JavaScript",
          "X-Fern-Runtime": core.RUNTIME.type,
          "X-Fern-Runtime-Version": core.RUNTIME.version,
        },
        contentType: "application/json",
        requestType: "json",
        body: serializers.CreateUserPool.jsonOrThrow(request, {
          unrecognizedObjectKeys: "strip",
        }),
        timeoutMs:
          (requestOptions === null || requestOptions === void 0
            ? void 0
            : requestOptions.timeoutInSeconds) != null
            ? requestOptions.timeoutInSeconds * 1000
            : 60000,
        maxRetries:
          requestOptions === null || requestOptions === void 0
            ? void 0
            : requestOptions.maxRetries,
        abortSignal:
          requestOptions === null || requestOptions === void 0
            ? void 0
            : requestOptions.abortSignal,
      });
      if (_response.ok) {
        return serializers.UserPool.parseOrThrow(_response.body, {
          unrecognizedObjectKeys: "passthrough",
          allowUnrecognizedUnionMembers: true,
          allowUnrecognizedEnumValues: true,
          breadcrumbsPrefix: ["response"],
        });
      }
      if (_response.error.reason === "status-code") {
        throw new errors.AnonApiError({
          statusCode: _response.error.statusCode,
          body: _response.error.body,
        });
      }
      switch (_response.error.reason) {
        case "non-json":
          throw new errors.AnonApiError({
            statusCode: _response.error.statusCode,
            body: _response.error.rawBody,
          });
        case "timeout":
          throw new errors.AnonApiTimeoutError();
        case "unknown":
          throw new errors.AnonApiError({
            message: _response.error.errorMessage,
          });
      }
    });
  }
  /**
   * This endpoint looks up an existing `UserPool`.
   *
   * @param {string} userPoolId - ID of the `UserPool`.
   * @param {UserPool.RequestOptions} requestOptions - Request-specific configuration.
   *
   * @example
   *     await client.userPool.getUserPool("string")
   */
  getUserPool(userPoolId, requestOptions) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
      const _response = yield core.fetcher({
        url: (0, url_join_1.default)(
          (_a = yield core.Supplier.get(this._options.environment)) !== null &&
            _a !== void 0
            ? _a
            : environments.AnonApiEnvironment.Sandbox,
          `/org/userPool/${encodeURIComponent(userPoolId)}`,
        ),
        method: "GET",
        headers: {
          Authorization: yield this._getAuthorizationHeader(),
          "X-Anon-Sdk-Client-Id": yield core.Supplier.get(
            this._options.anonSdkClientId,
          ),
          "X-Fern-Language": "JavaScript",
          "X-Fern-Runtime": core.RUNTIME.type,
          "X-Fern-Runtime-Version": core.RUNTIME.version,
        },
        contentType: "application/json",
        requestType: "json",
        timeoutMs:
          (requestOptions === null || requestOptions === void 0
            ? void 0
            : requestOptions.timeoutInSeconds) != null
            ? requestOptions.timeoutInSeconds * 1000
            : 60000,
        maxRetries:
          requestOptions === null || requestOptions === void 0
            ? void 0
            : requestOptions.maxRetries,
        abortSignal:
          requestOptions === null || requestOptions === void 0
            ? void 0
            : requestOptions.abortSignal,
      });
      if (_response.ok) {
        return serializers.UserPool.parseOrThrow(_response.body, {
          unrecognizedObjectKeys: "passthrough",
          allowUnrecognizedUnionMembers: true,
          allowUnrecognizedEnumValues: true,
          breadcrumbsPrefix: ["response"],
        });
      }
      if (_response.error.reason === "status-code") {
        throw new errors.AnonApiError({
          statusCode: _response.error.statusCode,
          body: _response.error.body,
        });
      }
      switch (_response.error.reason) {
        case "non-json":
          throw new errors.AnonApiError({
            statusCode: _response.error.statusCode,
            body: _response.error.rawBody,
          });
        case "timeout":
          throw new errors.AnonApiTimeoutError();
        case "unknown":
          throw new errors.AnonApiError({
            message: _response.error.errorMessage,
          });
      }
    });
  }
  /**
   * Lists all UserPools in your organization.
   *
   * @param {UserPool.RequestOptions} requestOptions - Request-specific configuration.
   *
   * @example
   *     await client.userPool.listUserPools()
   */
  listUserPools(requestOptions) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
      const _response = yield core.fetcher({
        url: (0, url_join_1.default)(
          (_a = yield core.Supplier.get(this._options.environment)) !== null &&
            _a !== void 0
            ? _a
            : environments.AnonApiEnvironment.Sandbox,
          "/org/userPools",
        ),
        method: "GET",
        headers: {
          Authorization: yield this._getAuthorizationHeader(),
          "X-Anon-Sdk-Client-Id": yield core.Supplier.get(
            this._options.anonSdkClientId,
          ),
          "X-Fern-Language": "JavaScript",
          "X-Fern-Runtime": core.RUNTIME.type,
          "X-Fern-Runtime-Version": core.RUNTIME.version,
        },
        contentType: "application/json",
        requestType: "json",
        timeoutMs:
          (requestOptions === null || requestOptions === void 0
            ? void 0
            : requestOptions.timeoutInSeconds) != null
            ? requestOptions.timeoutInSeconds * 1000
            : 60000,
        maxRetries:
          requestOptions === null || requestOptions === void 0
            ? void 0
            : requestOptions.maxRetries,
        abortSignal:
          requestOptions === null || requestOptions === void 0
            ? void 0
            : requestOptions.abortSignal,
      });
      if (_response.ok) {
        return serializers.UserPoolList.parseOrThrow(_response.body, {
          unrecognizedObjectKeys: "passthrough",
          allowUnrecognizedUnionMembers: true,
          allowUnrecognizedEnumValues: true,
          breadcrumbsPrefix: ["response"],
        });
      }
      if (_response.error.reason === "status-code") {
        throw new errors.AnonApiError({
          statusCode: _response.error.statusCode,
          body: _response.error.body,
        });
      }
      switch (_response.error.reason) {
        case "non-json":
          throw new errors.AnonApiError({
            statusCode: _response.error.statusCode,
            body: _response.error.rawBody,
          });
        case "timeout":
          throw new errors.AnonApiTimeoutError();
        case "unknown":
          throw new errors.AnonApiError({
            message: _response.error.errorMessage,
          });
      }
    });
  }
  _getAuthorizationHeader() {
    return __awaiter(this, void 0, void 0, function* () {
      return `Bearer ${yield core.Supplier.get(this._options.token)}`;
    });
  }
}
exports.UserPool = UserPool;
