// @anon/fern-sdk/dist/serialization/resources/index.js
"use strict";
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (
          !desc ||
          ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)
        ) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k];
            },
          };
        }
        Object.defineProperty(o, k2, desc);
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
      });
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
      }
    : function (o, v) {
        o["default"] = v;
      });
var __importStar =
  (this && this.__importStar) ||
  function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null)
      for (var k in mod)
        if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
          __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
  };
var __exportStar =
  (this && this.__exportStar) ||
  function (m, exports) {
    for (var p in m)
      if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p))
        __createBinding(exports, m, p);
  };
Object.defineProperty(exports, "__esModule", { value: true });
exports.userPool =
  exports.session =
  exports.sdkClient =
  exports.runtimeLog =
  exports.run =
  exports.organization =
  exports.member =
  exports.link =
  exports.appUserIdToken =
  exports.appIntegrations =
  exports.activity =
  exports.actions =
  exports.error =
    void 0;
exports.error = __importStar(require("./error"));
__exportStar(require("./error/errors"), exports);
exports.actions = __importStar(require("./actions"));
__exportStar(require("./actions/types"), exports);
exports.activity = __importStar(require("./activity"));
__exportStar(require("./activity/types"), exports);
exports.appIntegrations = __importStar(require("./appIntegrations"));
__exportStar(require("./appIntegrations/types"), exports);
exports.appUserIdToken = __importStar(require("./appUserIdToken"));
__exportStar(require("./appUserIdToken/types"), exports);
__exportStar(require("./error/types"), exports);
exports.link = __importStar(require("./link"));
__exportStar(require("./link/types"), exports);
exports.member = __importStar(require("./member"));
__exportStar(require("./member/types"), exports);
exports.organization = __importStar(require("./organization"));
__exportStar(require("./organization/types"), exports);
exports.run = __importStar(require("./run"));
__exportStar(require("./run/types"), exports);
exports.runtimeLog = __importStar(require("./runtimeLog"));
__exportStar(require("./runtimeLog/types"), exports);
exports.sdkClient = __importStar(require("./sdkClient"));
__exportStar(require("./sdkClient/types"), exports);
exports.session = __importStar(require("./session"));
__exportStar(require("./session/types"), exports);
exports.userPool = __importStar(require("./userPool"));
__exportStar(require("./userPool/types"), exports);
__exportStar(require("./session/client/requests"), exports);
