// @anon/account-types/dist/appIntegrationConfigRegistry.js
import { WebValidationMatchRule, } from "./appIntegrationConfig.js";
export const AppIntegrationConfigRegistry = {
    linkedin: {
        authUrl: "https://linkedin.com",
        displayName: "LinkedIn",
        iconUrl: "https://logo.clearbit.com/linkedin.com",
        completionIndicators: [
            {
                storage: "cookie",
                key: "li_at",
                domains: [".www.linkedin.com"],
            },
        ],
        sessionExtractionDomains: [
            {
                domain: ".linkedin.com",
            },
            {
                domain: ".www.linkedin.com",
            },
        ],
        sessionValidation: {
            sessionValidityChecks: [
                {
                    url: "https://www.linkedin.com/mypreferences/d/categories/account",
                    criteria: [
                        {
                            selector: 'h2:has-text("General preferences")',
                            exists: true,
                        },
                    ],
                    matchRule: WebValidationMatchRule.ANY,
                },
            ],
        },
    },
    facebook: {
        authUrl: "https://facebook.com",
        displayName: "Facebook",
        iconUrl: "https://logo.clearbit.com/facebook.com",
        domCompletionIndicator: `//*[contains(@aria-label, 'Create story')]`,
        completionIndicators: [
            {
                storage: "cookie",
                key: "c_user",
                domains: [".facebook.com"],
            },
            {
                storage: "cookie",
                key: "sb",
                domains: [".facebook.com"],
            },
            {
                storage: "cookie",
                key: "xs",
                domains: [".facebook.com"],
            },
        ],
        sessionExtractionDomains: [
            {
                domain: ".facebook.com",
            },
            {
                domain: "www.facebook.com",
            },
        ],
        sessionValidation: {
            sessionValidityChecks: [],
        },
    },
    github: {
        authUrl: "https://github.com/login",
        displayName: "Github",
        iconUrl: "https://logo.clearbit.com/github.com",
        completionIndicators: [
            {
                storage: "cookie",
                key: "user_session",
                domains: ["github.com"],
            },
        ],
        sessionExtractionDomains: [
            {
                domain: "github.com",
            },
            {
                domain: ".github.com",
            },
        ],
        sessionValidation: {
            sessionValidityChecks: [
                {
                    url: "https://github.com/settings/profile",
                    criteria: [
                        {
                            selector: 'h2:has-text("Public profile")',
                            exists: true,
                        },
                    ],
                    matchRule: WebValidationMatchRule.ANY,
                },
            ],
        },
    },
    instagram: {
        authUrl: "https://www.instagram.com/accounts/login/",
        displayName: "Instagram",
        iconUrl: "https://logo.clearbit.com/instagram.com",
        completionIndicators: [
            {
                storage: "cookie",
                key: "sessionid",
                domains: [".instagram.com"],
            },
            {
                storage: "cookie",
                key: "ig_did",
                domains: [".instagram.com", "www.instagram.com"],
            },
        ],
        sessionExtractionDomains: [
            {
                domain: ".instagram.com",
            },
            {
                domain: ".facebook.com",
            },
        ],
        sessionValidation: {
            sessionValidityChecks: [],
        },
    },
    uber: {
        authUrl: "https://m.uber.com/",
        displayName: "Uber",
        iconUrl: "https://logo.clearbit.com/uber.com",
        domCompletionIndicator: "//h1[text()='Get a ride']",
        completionIndicators: [
            {
                storage: "cookie",
                key: "jwt-session",
                domains: ["m.uber.com"],
            },
        ],
        sessionExtractionDomains: [
            {
                domain: "www.uber.com",
            },
            {
                domain: ".uber.com",
            },
            {
                domain: "m.uber.com",
            },
            {
                domain: "auth.uber.com",
            },
            {
                domain: ".auth.uber.com",
            },
            {
                domain: ".www.uber.com",
            },
            {
                domain: ".payments.uber.com",
            },
            {
                domain: "payments.uber.com",
            },
        ],
        sessionValidation: {
            sessionValidityChecks: [
                {
                    url: "https://account.uber.com/",
                    criteria: [
                        {
                            selector: 'span:has-text("For Me")',
                            exists: true,
                        },
                    ],
                    matchRule: WebValidationMatchRule.ANY,
                },
            ],
        },
    },
    uber_eats: {
        experimental: true,
        authUrl: "https://www.ubereats.com/login-redirect",
        displayName: "Uber Eats",
        iconUrl: "https://logo.clearbit.com/ubereats.com",
        domCompletionIndicator: "//button[@aria-label='Carts']",
        completionIndicators: [
            {
                storage: "cookie",
                key: "jwt-session",
                domains: ["www.ubereats.com"],
            },
        ],
        sessionExtractionDomains: [
            {
                domain: "www.ubereats.com",
            },
            {
                domain: ".ubereats.com",
            },
            {
                domain: "ubereats.com",
            },
            {
                domain: "www.uber.com",
            },
            {
                domain: ".uber.com",
            },
            {
                domain: "m.uber.com",
            },
            {
                domain: "auth.uber.com",
            },
            {
                domain: ".auth.uber.com",
            },
            {
                domain: ".www.uber.com",
            },
            {
                domain: ".payments.uber.com",
            },
            {
                domain: "payments.uber.com",
            },
        ],
        sessionValidation: {
            sessionValidityChecks: [
                {
                    url: "https://account.uber.com/",
                    criteria: [
                        {
                            selector: 'h2:has-text("Basic Info")',
                            exists: true,
                        },
                    ],
                    matchRule: WebValidationMatchRule.ANY,
                },
            ],
        },
    },
    delta: {
        authUrl: "https://www.delta.com/skymiles/login",
        displayName: "Delta Air Lines",
        iconUrl: "https://logo.clearbit.com/delta.com",
        completionIndicators: [
            {
                storage: "cookie",
                key: "isin",
                domains: [".delta.com", "www.delta.com"],
            },
        ],
        sessionExtractionDomains: [
            {
                domain: ".delta.com",
            },
            {
                domain: "www.delta.com",
            },
            {
                domain: "signin.delta.com",
            },
            {
                domain: ".dotomi.com",
            },
            {
                domain: ".adnxs.com",
            },
            {
                domain: ".demdex.net",
            },
        ],
        sessionValidation: {
            sessionValidityChecks: [
                {
                    url: "https://www.delta.com/myprofile/personal-details",
                    criteria: [
                        {
                            selector: 'h3:has-text("Basic Info")',
                            exists: true,
                        },
                    ],
                    matchRule: WebValidationMatchRule.ANY,
                },
            ],
        },
    },
    united_airlines: {
        authUrl: "https://www.united.com/en/us/united-mileageplus-signin/",
        displayName: "United Airlines",
        iconUrl: "https://logo.clearbit.com/united.com",
        completionIndicators: [
            {
                storage: "cookie",
                key: "Session",
                domains: [".united.com", "www.united.com"],
            },
            {
                storage: "cookie",
                key: "User",
                domains: [".united.com", "www.united.com"],
            },
        ],
        sessionExtractionDomains: [
            {
                domain: ".united.com",
            },
            {
                domain: "www.united.com",
            },
        ],
        sessionValidation: {
            sessionValidityChecks: [
            // {
            //note(boateng): Is the /en/us path too strict? Should atleast still work for usa customers
            //FIXME(boateng): United seems to be blocking when performing session validity checks, i.e injecting the session
            //   url: "https://www.united.com/en/us/myunited",
            //   criteria: [
            //     {
            //       selector: "text=/My United/",
            //       exists: true,
            //     },
            //   ],
            //   matchRule: WebValidationMatchRule.ANY,
            // },
            ],
        },
    },
    instacart: {
        displayName: "Instacart",
        iconUrl: "https://instacart.com/favicon.ico",
        authUrl: "https://www.instacart.com/login?next=%2Fstore",
        completionIndicators: [
            {
                storage: "cookie",
                key: "__Host-instacart_sid",
                domains: ["www.instacart.com", ".instacart.com"],
            },
        ],
        sessionExtractionDomains: [
            {
                domain: ".instacart.com",
            },
            {
                domain: "www.instacart.com",
            },
            {
                domain: ".www.instacart.com",
            },
        ],
        sessionValidation: {
            sessionValidityChecks: [
                {
                    url: "https://www.instacart.com/store/account",
                    criteria: [
                        {
                            selector: "text=/Account information/",
                            exists: true,
                        },
                    ],
                    matchRule: WebValidationMatchRule.ANY,
                },
            ],
        },
    },
    resy: {
        displayName: "Resy",
        iconUrl: "https://resy.com/favicon.ico",
        authUrl: "https://resy.com/account/profile",
        domCompletionIndicator: "//div[contains(@class, 'logged-in')]",
        completionIndicators: [
            {
                storage: "cookie",
                key: "authToken",
                domains: ["resy.com", ".resy.com"],
            },
        ],
        sessionExtractionDomains: [
            {
                domain: ".resy.com",
            },
            {
                domain: "www.resy.com",
            },
            {
                domain: "resy.com",
            },
        ],
        sessionValidation: {
            sessionValidityChecks: [
                {
                    url: "https://resy.com/account/profile",
                    criteria: [
                        {
                            selector: 'h2:has-text("Profile")',
                            exists: true,
                        },
                    ],
                    matchRule: WebValidationMatchRule.ANY,
                },
            ],
        },
    },
    netflix: {
        displayName: "Netflix",
        iconUrl: "https://netflix.com/favicon.ico",
        authUrl: "https://www.netflix.com/login",
        domCompletionIndicator: "//div[contains(@class, 'title-boxart')]",
        completionIndicators: [
            {
                storage: "cookie",
                key: "NetflixId",
                domains: ["netflix.com", ".netflix.com", "www.netflix.com"],
            },
            {
                storage: "cookie",
                key: "SecureNetflixId",
                domains: ["netflix.com", ".netflix.com", "www.netflix.com"],
            },
        ],
        sessionExtractionDomains: [
            {
                domain: ".netflix.com",
            },
            {
                domain: "www.netflix.com",
            },
            {
                domain: "netflix.com",
            },
        ],
        sessionValidation: {
            sessionValidityChecks: [
                {
                    url: "https://www.netflix.com/account",
                    criteria: [
                        {
                            selector: 'h2:has-text("Profiles")',
                            exists: true,
                        },
                    ],
                    matchRule: WebValidationMatchRule.ANY,
                },
            ],
        },
    },
    amazon: {
        displayName: "Amazon",
        iconUrl: "https://amazon.com/favicon.ico",
        authUrl: "https://www.amazon.com/gp/sign-in.html",
        domCompletionIndicator: "//div[contains(@id, 'nav-global-location-slot')]",
        completionIndicators: [
            {
                storage: "cookie",
                key: "sess-at-main",
                domains: [".amazon.com"],
            },
        ],
        sessionExtractionDomains: [
            {
                domain: ".amazon.com",
            },
            {
                domain: "www.amazon.com",
            },
        ],
        sessionValidation: {
            sessionValidityChecks: [
                {
                    url: "https://www.amazon.com/gp/your-account/order-history",
                    criteria: [
                        {
                            selector: 'h1:has-text("Your Orders")',
                            exists: true,
                        },
                    ],
                    matchRule: WebValidationMatchRule.ANY,
                },
            ],
        },
    },
    opentable: {
        displayName: "OpenTable",
        iconUrl: "https://opentable.com/favicon.ico",
        authUrl: "https://www.opentable.com/authenticate/start?isPopup=false",
        domCompletionIndicator: "//h2[text()='Points']",
        completionIndicators: [
            {
                storage: "cookie",
                key: "authCke",
                domains: [".opentable.com", "www.opentable.com"],
            },
        ],
        sessionExtractionDomains: [
            {
                domain: ".opentable.com",
            },
            {
                domain: "www.opentable.com",
            },
        ],
        sessionValidation: {
            sessionValidityChecks: [
                {
                    url: "https://www.opentable.com/user/dining-dashboard",
                    criteria: [
                        {
                            selector: 'h2:has-text("Points")',
                            exists: true,
                        },
                    ],
                    matchRule: WebValidationMatchRule.ANY,
                },
            ],
        },
    },
    twitter: {
        authUrl: "https://twitter.com",
        displayName: "Twitter",
        iconUrl: "https://abs.twimg.com/favicons/twitter.3.ico",
        completionIndicators: [
            {
                storage: "cookie",
                key: "auth_token",
                domains: [".twitter.com", "www.twitter.com", ".x.com", "www.x.com"],
            },
            {
                storage: "cookie",
                key: "twid",
                domains: [".twitter.com", "www.twitter.com", ".x.com", "www.x.com"],
            },
        ],
        sessionExtractionDomains: [
            {
                domain: ".twitter.com",
            },
            {
                domain: ".x.com",
            },
        ],
        sessionValidation: {
            sessionValidityChecks: [
                {
                    url: "https://x.com/settings/account",
                    criteria: [
                        {
                            selector: 'span:has-text("Account Information")',
                            exists: true,
                        },
                        {
                            selector: 'span:has-text("Your account")',
                            exists: true,
                        },
                    ],
                    matchRule: WebValidationMatchRule.ANY,
                },
            ],
        },
    },
    doordash: {
        authUrl: "https://www.doordash.com/orders/",
        displayName: "Doordash",
        iconUrl: "https://logo.clearbit.com/doordash.com",
        sessionExtractionDomains: [
            {
                domain: ".doordash.com",
            },
            {
                domain: "www.doordash.com",
            },
        ],
        domCompletionIndicator: "//span[text()='Orders']",
        sessionValidation: {
            sessionValidityChecks: [
                {
                    url: "https://www.doordash.com/consumer/edit_profile/",
                    criteria: [
                        {
                            selector: 'h2:has-text("Profile")',
                            exists: true,
                        },
                    ],
                    matchRule: WebValidationMatchRule.ANY,
                },
            ],
        },
        completionIndicators: [
            {
                storage: "cookie",
                key: "ddweb_token",
                domains: [".doordash.com", "www.doordash.com"],
            },
            {
                storage: "cookie",
                key: "ddweb_session_id",
                domains: [".doordash.com", "www.doordash.com"],
            },
        ],
    },
    google: {
        experimental: true,
        authUrl: "https://accounts.google.com/",
        displayName: "Google",
        iconUrl: "https://logo.clearbit.com/google.com",
        sessionExtractionDomains: [
            {
                domain: ".google.com",
            },
            {
                domain: "www.google.com",
            },
            {
                domain: "accounts.google.com",
            },
            {
                domain: ".ogs.google.com",
            },
            {
                domain: "myactivity.google.com",
            },
            {
                domain: "mail.google.com",
            },
            {
                domain: "chat.google.com",
            },
            {
                domain: "contacts.google.com",
            },
            {
                domain: "drive.google.com",
            },
            {
                domain: "docs.google.com",
            },
        ],
        domCompletionIndicator: "//div[text()='Home']",
        sessionValidation: {
            sessionValidityChecks: [
                {
                    url: "https://myaccount.google.com/",
                    criteria: [
                        {
                            selector: "//div[text()='Home']",
                            exists: true,
                        },
                    ],
                    matchRule: WebValidationMatchRule.ANY,
                },
            ],
        },
        completionIndicators: [
            {
                storage: "cookie",
                key: "__Secure-1PSIDCC",
                domains: [".google.com", "accounts.google.com"],
            },
        ],
    },
};
