// @anon/account-types/dist/notifications.js
import { z } from "zod";
export const iOSPushNotificationRequestSchema = z.object({
    iosPushAuthorization: z.object({
        iosAppId: z.string(),
        authToken: z.string(),
    }),
    account: z.object({
        ownerId: z.string(),
        domain: z.string(),
    }),
    appSdkClientId: z.string(),
    alertBody: z.string().optional(),
    alertTitle: z.string().optional(),
});
// FIXME: this is currently unused by downstream sdks and account-svc doesn't handle it
export const ChromePushNotificationRequestSchema = z.object({
    appUserId: z.string(),
    app: z.string(),
});
export const PushNotificationRequestSchema = iOSPushNotificationRequestSchema.or(ChromePushNotificationRequestSchema);
export const iOSPushNotificationPayloadSchema = z.object({
    account: z.object({
        ownerId: z.string(),
        domain: z.string(),
    }),
    appSdkClientId: z.string(),
    iosAppId: z.string(),
    authenticationToken: z.string(),
    alertTitle: z.string().optional(),
    alertBody: z.string().optional(),
    //pushType: z.string().optional()
});
export const ChromePushNotificationPayloadSchema = z.object({
    appUserId: z.string(),
    app: z.string(),
});
export const PushNotificationErrorDetailsSchema = z.object({
    appUserId: z.string(),
    appSdkClientId: z.string(),
    domain: z.string().optional(),
    orgSvcStatusCode: z.number().optional(),
});
export const PushNotificationResponseSchema = z.object({
    succeeded: z.boolean(),
    appUserId: z.string(),
    appSdkClientId: z.string(),
    domain: z.string().optional(),
});
