// @anon/fern-sdk/dist/serialization/resources/sdkClient/types/SdkClient.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (
          !desc ||
          ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)
        ) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k];
            },
          };
        }
        Object.defineProperty(o, k2, desc);
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
      });
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
      }
    : function (o, v) {
        o["default"] = v;
      });
var __importStar =
  (this && this.__importStar) ||
  function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null)
      for (var k in mod)
        if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
          __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
  };
Object.defineProperty(exports, "__esModule", { value: true });
exports.SdkClient = void 0;
const core = __importStar(require("../../../../core"));
const SdkClientId_1 = require("./SdkClientId");
const OrganizationId_1 = require("../../organization/types/OrganizationId");
const SdkAuth_1 = require("./SdkAuth");
exports.SdkClient = core.serialization.object({
  id: SdkClientId_1.SdkClientId,
  name: core.serialization.string(),
  description: core.serialization.string(),
  organizationId: OrganizationId_1.OrganizationId,
  auth: SdkAuth_1.SdkAuth,
});
