// @anon/fern-sdk/dist/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnonApiClient = void 0;
const Client_1 = require("./api/resources/actions/client/Client");
const Client_2 = require("./api/resources/activity/client/Client");
const Client_3 = require("./api/resources/appIntegrations/client/Client");
const Client_4 = require("./api/resources/appUserIdToken/client/Client");
const Client_5 = require("./api/resources/link/client/Client");
const Client_6 = require("./api/resources/member/client/Client");
const Client_7 = require("./api/resources/organization/client/Client");
const Client_8 = require("./api/resources/run/client/Client");
const Client_9 = require("./api/resources/runtimeLog/client/Client");
const Client_10 = require("./api/resources/sdkClient/client/Client");
const Client_11 = require("./api/resources/session/client/Client");
const Client_12 = require("./api/resources/userPool/client/Client");
class AnonApiClient {
  constructor(_options) {
    this._options = _options;
  }
  get actions() {
    var _a;
    return (_a = this._actions) !== null && _a !== void 0
      ? _a
      : (this._actions = new Client_1.Actions(this._options));
  }
  get activity() {
    var _a;
    return (_a = this._activity) !== null && _a !== void 0
      ? _a
      : (this._activity = new Client_2.Activity(this._options));
  }
  get appIntegrations() {
    var _a;
    return (_a = this._appIntegrations) !== null && _a !== void 0
      ? _a
      : (this._appIntegrations = new Client_3.AppIntegrations(this._options));
  }
  get appUserIdToken() {
    var _a;
    return (_a = this._appUserIdToken) !== null && _a !== void 0
      ? _a
      : (this._appUserIdToken = new Client_4.AppUserIdToken(this._options));
  }
  get link() {
    var _a;
    return (_a = this._link) !== null && _a !== void 0
      ? _a
      : (this._link = new Client_5.Link(this._options));
  }
  get member() {
    var _a;
    return (_a = this._member) !== null && _a !== void 0
      ? _a
      : (this._member = new Client_6.Member(this._options));
  }
  get organization() {
    var _a;
    return (_a = this._organization) !== null && _a !== void 0
      ? _a
      : (this._organization = new Client_7.Organization(this._options));
  }
  get run() {
    var _a;
    return (_a = this._run) !== null && _a !== void 0
      ? _a
      : (this._run = new Client_8.Run(this._options));
  }
  get runtimeLog() {
    var _a;
    return (_a = this._runtimeLog) !== null && _a !== void 0
      ? _a
      : (this._runtimeLog = new Client_9.RuntimeLog(this._options));
  }
  get sdkClient() {
    var _a;
    return (_a = this._sdkClient) !== null && _a !== void 0
      ? _a
      : (this._sdkClient = new Client_10.SdkClient(this._options));
  }
  get session() {
    var _a;
    return (_a = this._session) !== null && _a !== void 0
      ? _a
      : (this._session = new Client_11.Session(this._options));
  }
  get userPool() {
    var _a;
    return (_a = this._userPool) !== null && _a !== void 0
      ? _a
      : (this._userPool = new Client_12.UserPool(this._options));
  }
}
exports.AnonApiClient = AnonApiClient;
